geochem_inverse_optimize.py

```diff
@@ -75,56 +75,148 @@ def get_sample_graphs(data_dir: str) -> Tuple[nx.DiGraph, Any]:
+            self._primary_terms.append(cp_log_ratio_norm(normalised_concentration, observed))
+
+            if ds := nx_get_downstream(self.sample_network, sample_name):
+                downstream_data = self.sample_network.nodes[ds]["data"]
+                # Add our flux to the downstream node's
+                downstream_data.total_flux += my_data.total_flux
+
+    def _build_regularizer_terms(self) -> None:
+        # Build the regularizer
+        sample_names = [
+            node for node in self.sample_network.nodes
+        ]  # Sample names for each node. TODO. MAKE THIS A DICTIONARY PART OF THE CLASS
+        for adjacent_nodes, border_length in self.sample_adjacency.items():
+            # NB adjacency network stores samples as numbers from 1 - 63 (*not* 0 - 62)
+            node_a = sample_names[adjacent_nodes[0] - 1]  # Get samplename from node 'number'
+            node_b = sample_names[adjacent_nodes[1] - 1]
+            a_concen = self.sample_network.nodes[node_a]["data"].my_value
+            b_concen = self.sample_network.nodes[node_b]["data"].my_value
+            #           self._regularizer_terms.append(border_length * (cp_log_ratio_norm(a_concen,b_concen)))  # < 'log-ratio' difference (preferable)
+            self._regularizer_terms.append(
+                border_length * (a_concen - b_concen)
+            )  # simple difference (not desirable)
+
+    def _build_problem(self) -> None:
+        assert self._primary_terms
+
+        if not self._regularizer_terms:
+            print("WARNING: No regularizer terms found!")
+
+        # Build the objective and constraints
+        objective = cp.norm(cp.vstack(self._primary_terms))
+        if self._regularizer_terms:
+            # TODO(alexlipp,r-barnes): Make sure that his uses the same summation strategy as the primary terms
+            objective += self._regularizer_strength * cp.norm(cp.vstack(self._regularizer_terms))
+        constraints = []
+
+        # Create and solve the problem
+        print("Compiling problem...")
+        self._problem = cp.Problem(cp.Minimize(objective), constraints)
+
+    def solve(
+        self,
+        observation_data: ElementData,
+        regularization_strength: Optional[float] = None,
+        solver: str = "gurobi",
+    ):
+        obs_mean = np.mean(list(observation_data.values()))
+
+        # Reset all sites' observations
+        for x in self._site_to_parameter.values():
+            x.value = None
+        # Assign each observed value to a site, making sure that the site exists
+        for site, value in observation_data.items():
+            assert site in self._site_to_parameter
+            self._site_to_parameter[site].value = value / obs_mean
+        # Ensure that all sites in the problem were assigned
+        for x in self._site_to_parameter.values():
+            assert x.value is not None
+
+        if self._regularizer_terms and not regularization_strength:
+            print(
```

Comment:
Can you please refactor this so it uses
```
raise Exception("...")
```
if regularizer terms are present but no strength is assigned. I'd prefer not to set random defaults and instead inform the user that they need to make choices.

This should also eliminate the need for the `else`.